File: web/js/peacoq-ts/src/feedback-content.ts
export abstract class FeedbackContent {
  static create(f: { tag: string; contents: Object }) {
    switch (f.tag) {
      case "AddedAxiom":
        return new AddedAxiom();
      case "Custom":
        console.log("TODO: FeedbackContent for " + f.tag, f);
        break;
      case "ErrorMsg":
        return new ErrorMsg(f.contents);
      case "FileDependency":
        return new FileDependency(f.contents);
      case "FileLoaded":
        return new FileLoaded(f.contents);
      case "GlobDef":
      case "GlobRef":
      case "Goals":
      case "Message":
        console.log("TODO: FeedbackContent for " + f.tag, f);
        break;
      case "Processed":
        return new Processed();
      case "ProcessingIn":
        return new ProcessingIn(f.contents);
      case "WorkerStatus":
        console.log("TODO: FeedbackContent for " + f.tag, f);
        break;
      // other tags don't need fields
      default:
        throw ("Unknown FeedbackContent tag: " + f.tag);
    }
  }
}

export class AddedAxiom extends FeedbackContent { }

export class ErrorMsg extends FeedbackContent {
  message: string;
  start: number;
  stop: number;
  constructor(c) {
    super();
    let [[start, stop], message] = c;
    this.start = start;
    this.stop = stop;
    this.message = replaceNBSPWithSpaces(message);
  }
}

export class FileDependency extends FeedbackContent {
  dependsOnFile: string;
  file: string;
  constructor(c) {
    super();
    let [file, dependsOnFile] = c;
    this.dependsOnFile = dependsOnFile;
    this.file = file;
  }
}

export class FileLoaded extends FeedbackContent {
  path: string;
  qualifiedModuleName: string;
  constructor(c) {
    super();
    let [qualifiedModuleName, path] = c;
    this.path = path;
    this.qualifiedModuleName = qualifiedModuleName;
  }
}

export class Processed extends FeedbackContent {
  toString() { return "Processed"; }
}

export class ProcessingIn extends FeedbackContent {
  s: string;
  constructor(s) {
    super();
    this.s = s;
  }
  toString() {
    return "ProcessingIn(" + this.s + ")";
  }
}

File: peacoqtop/lib/Coq/FeedbackContent.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DeriveGeneric #-}

module Coq.FeedbackContent where

import Data.Aeson            (ToJSON)
import Data.Proxy
import GHC.Generics          (Generic)
import Text.XML.Stream.Parse (tagName)

import Coq.Loc
import Coq.Message
import FromXML
import Utils

data FeedbackContent
  = AddedAxiom
  | Complete
  | Custom Loc String () -- TODO: xml
  | ErrorMsg Loc String
  | FileDependency (Maybe String) String
  | FileLoaded String String
  | GlobDef Loc String String String
  | GlobRef Loc String String String String
  | Goals Loc String
  | Incomplete
  | InProgress
  | Message Message
  | Processed
  | ProcessingIn String
  | WorkerStatus String String
  deriving (Generic, Show)

instance ToJSON FeedbackContent

instance FromXML FeedbackContent where
  instanceName Proxy = "FeedbackContent"
  parseXML = tagName "feedback_content" (unpackRequireAttr "val") $ \ val -> do
    case val of
      "addedaxiom" ->
        return AddedAxiom
      "complete" ->
        return Complete
      "errormsg" -> do
        loc <- forceXML
        err <- forceXML
        return $ ErrorMsg loc err
      "filedependency" -> do
        from <- forceXML
        dep <- forceXML
        return $ FileDependency from dep
      "fileloaded" -> do
        dirPath <- forceXML
        fileName <- forceXML
        return $ FileLoaded dirPath fileName
      "processed" ->
        return Processed
      "processingin" ->
        ProcessingIn <$> forceXML
      "workerstatus" -> do
        workerId <- forceXML
        status <- forceXML
        return $ WorkerStatus workerId status
      _ -> error $ "Unknown feedback_content: " ++ val
